lib/streamlit/runtime/forward_msg_cache.py

```diff
@@ -39,27 +32,38 @@ def populate_hash_if_needed(msg: ForwardMsg) -> str:
     ----------
     msg : ForwardMsg
 
-    Returns
-    -------
-    string
-        The message's hash, returned here for convenience. (The hash
-        will also be assigned to the ForwardMsg; callers do not need
-        to do this.)
-
     """
-    if msg.hash == "":
+    if msg.hash == "" and msg.WhichOneof("type") not in {"ref_hash", "initialize"}:
         # Move the message's metadata aside. It's not part of the
         # hash calculation.
         metadata = msg.metadata
         msg.ClearField("metadata")
 
+        # Serialize the message to bytes using the deterministic serializer to
+        # ensure consistent hashing.
+        serialized_msg = msg.SerializeToString(deterministic=True)
+
+        # TODO(lukasmasuch): Evaluate more optimized hashing for larger messages:
+        # - Add the type element type and number of bytes to the hash.
+        # - Only hash the first N bytes of the message.
+
         # MD5 is good enough for what we need, which is uniqueness.
-        msg.hash = util.calc_md5(msg.SerializeToString())
+        msg.hash = util.calc_md5(serialized_msg)
 
         # Restore metadata.
         msg.metadata.CopyFrom(metadata)
 
-    return msg.hash
+        # Set cacheable flag if above the min cached size and if its a `new_element`
+        # delta. We only cache new_element and add_block deltas since container's
+        # are not expected to be larger than a few KB and have other side-effects
+        # to consider if cached. But `add_block` deltas should still get a hash.
+        # In case we ever allow other delta types to be cached, we should
+        # also need to adapt the composable logic in forward_msg_queue.
+        msg.metadata.cacheable = (
+            len(serialized_msg) >= int(config.get_option("global.minCachedMessageSize"))
```

Comment:
Related to [this comment](https://github.com/streamlit/streamlit/pull/9562/files#r2022409023), I am wondering whether the `minCachedMessage` property will still be necessary / still makes sense. I _assume_ in the past it was mainly to prevent storing too many messages on the server-side but now that we add a hash to every message anyways, this should perhaps be something that simply the frontend decides (by sending the cached hashes or not)

Reply:
Yep, agree. I think this is something we should investigate to simplify/clean up as a follow-up. I added this to [this tech debt item](https://www.notion.so/snowflake-corp/Let-the-frontend-decide-on-which-messages-to-cache-1c87170bb416800e8034e6531c62ca47) as well.